Argo/Globals/StandardTypes.swift

```diff
@@ -2,53 +2,64 @@ import Foundation
 import Runes
 
 extension String: JSONDecodable {
-  public static func decode(j: JSON) -> String? {
+  public static func fromJSON(j: JSON) -> Parser<String> {
     switch j {
-    case let .String(s): return s
-    default: return .None
+    case let .String(s): return pure(s)
+    default: return typeMismatch("String", j)
     }
   }
 }
 
 extension Int: JSONDecodable {
-  public static func decode(j: JSON) -> Int? {
+  public static func fromJSON(j: JSON) -> Parser<Int> {
     switch j {
-    case let .Number(n): return n as Int
-    default: return .None
+    case let .Number(n): return pure(n as Int)
+    default: return typeMismatch("Int", j)
     }
   }
 }
 
 extension Double: JSONDecodable {
-  public static func decode(j: JSON) -> Double? {
+  public static func fromJSON(j: JSON) -> Parser<Double> {
     switch j {
-    case let .Number(n): return n as Double
-    default: return .None
+    case let .Number(n): return pure(n as Double)
+    default: return typeMismatch("Double", j)
     }
   }
 }
 
 extension Bool: JSONDecodable {
-  public static func decode(j: JSON) -> Bool? {
+  public static func fromJSON(j: JSON) -> Parser<Bool> {
     switch j {
-    case let .Number(n): return n as Bool
-    default: return .None
+    case let .Number(n): return pure(n as Bool)
+    default: return typeMismatch("Bool", j)
     }
   }
 }
 
 extension Float: JSONDecodable {
-  public static func decode(j: JSON) -> Float? {
+  public static func fromJSON(j: JSON) -> Parser<Float> {
     switch j {
-    case let .Number(n): return n as Float
-    default: return .None
+    case let .Number(n): return pure(n as Float)
+    default: return typeMismatch("Float", j)
     }
   }
 }
 
-public func decodeArray<A where A: JSONDecodable, A == A.DecodedType>(value: JSON) -> [A]? {
+public func decodeArray<A where A: JSONDecodable, A == A.DecodedType>(value: JSON) -> Parser<[A]> {
   switch value {
-  case let .Array(a): return sequence({ A.decode($0) } <^> a)
-  default: return .None
+  case let .Array(a): return sequence({ A.fromJSON($0) } <^> a)
+  default: return typeMismatch("Array", value)
   }
 }
+
+public func decodeObject<A where A: JSONDecodable, A == A.DecodedType>(value: JSON) -> Parser<[String: A]> {
+  switch value {
+  case let .Object(o): return sequence(map(o) { A.fromJSON($0) })
```

Comment:
Is it weird that this is now using a different syntax than `decodeArray`, which is doing the exact same thing? Maybe we should change `decodeArray` to match this instead of defining `<^>` for dictionaries?